street_network/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,128 @@
-                    } else {
-                        _num_roads_out += 1
-                    }
-                }
-                None => num_roads_inout += 1,
+    // Calculate all the possible movements, except (U-turns).
+    //FIXME assert!(roads is sorted clockwise), which it isn't
+    let mut connections = Vec::new();
+    // Consider turns pairs of roads, from s to d, using their position as index, so we can them later.
+    for s in 0..road_ids.len() {
+        for d in 0..road_ids.len() {
+            if s == d {
+                continue;
+            }
+            // FIXME check if we can travel into the intersection on s and out of the intersection on d
+            if turn_is_allowed(roads.get(s).unwrap(), road_ids.get(d).unwrap()) {
+                connections.push((s, d));
             }
         }
+    }
 
-        if num_roads_inout == 0 {
-            // The simplest MultiConnect
-            return MultiConnection;
+    // Calculate all the collisions.
+    let mut conflicts = BTreeMap::new();
+    let mut worst_conflict = Uncontested;
+    // Compare every pair of connections. Use the order of the roads around the intersection to
+    // detect if they diverge, merge, or cross.
+    // assert!(connections is sorted) so small_con large_con makes sense.
+    let mut each_con = connections.iter();
+    while let Some(small_con) = each_con.next() {
+        for large_con in each_con.clone() {
+            let conflict = calc_conflict(small_con, large_con);
+            worst_conflict = max(worst_conflict, conflict);
+            conflicts.insert((small_con, large_con), conflict);
         }
+    }
+
+    match worst_conflict {
+        Cross => (Crossing, Cross),
+        c => (MultiConnection, c),
+    }
+}
 
-        // detect the simple case of a single road splitting
-        if num_roads_inout == 1 {
-            // TODO Determine if it is possible to turn from the 'in' road to the 'out' road.
-            // If not, then we have a dual carriageway split.
-            if false {
-                return MultiConnection;
+fn turn_is_allowed(src: &Road, dst_id: &OriginalRoad) -> bool {
+    let mut has_exclusive_allows = false;
+    for (t, other) in &src.turn_restrictions {
+        match t {
+            RestrictionType::BanTurns => {
+                if other == dst_id {
+                    return false;
+                }
+            }
+            RestrictionType::OnlyAllowTurns => {
+                if other == dst_id {
+                    return true;
+                }
+                has_exclusive_allows = true;
             }
         }
     }
+    !has_exclusive_allows
+}
 
-    // A Merge is characterised by an uninterrupted road (that would qualify as a Connection or
-    // MultiConnection), with additional connected roads that yield.
-    // TODO Combine roads into corridors and count them.
+fn calc_conflict(a: &(usize, usize), b: &(usize, usize)) -> ConflictType {
```

Comment:
The types are getting a little weird here, so I think passing around `&Road` pointers would indeed more clear. These're connections between two roads, and we're just looking to see if identity of the roads is the same

Reply:
Edit: never mind, `is_between` is using the ordering of the roads around each other. I understand now

Reply:
```rust
type Position = usize;
type ConnectedPositions = (usize, usize);
```
?

Reply:
Eh, if you want. If the `usize`s leaked outside this file, a typedef at least would be nice. I think a single comment near the top of this method could kind of explain what we're working with here, and that would suffice